src/utils/untar.js

```diff
@@ -0,0 +1,316 @@
+    const uid = parseInt(stream.readString(8), 10);
+    const gid = parseInt(stream.readString(8), 10);
+    const size = parseInt(stream.readString(12), 8);
+    const mtime = parseInt(stream.readString(12), 8);
+    const checksum = parseInt(stream.readString(8), 10);
+    const type = stream.readString(1);
+    const linkname = stream.readString(100);
+    const ustarFormat = stream.readString(6);
+    let version, uname, gname, devmajor, devminor, namePrefix;
+
+    if (ustarFormat.indexOf('ustar') > -1) {
+      version = stream.readString(2);
+      uname = stream.readString(32);
+      gname = stream.readString(32);
+      devmajor = parseInt(stream.readString(8), 10);
+      devminor = parseInt(stream.readString(8), 10);
+      namePrefix = stream.readString(155);
+
+      if (namePrefix.length > 0) {
+        name = namePrefix + '/' + name;
+      }
+    }
+
+    stream.seekTo(dataBeginPos);
+
+    let buffer = new Uint8Array(0);
+
+    // Derived from https://www.mkssoftware.com/docs/man4/pax.4.asp
+    // and https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.bpxa500/pxarchfm.htm
+    switch (type) {
+      case '0': // Normal file is either "0" or "\0".
+      case '': // In case of "\0", readString returns an empty string, that is "".
+        buffer = stream.readBuffer(size);
+        break;
+      case '1': // Link to another file already archived
+        // TODO Should we do anything with these?
+        break;
+      case '2': // Symbolic link
+        // TODO Should we do anything with these?
+        break;
+      case '3': // Character special device (what does this mean??)
+        break;
+      case '4': // Block special device
+        break;
+      case '5': // Directory
+        break;
+      case '6': // FIFO special file
+        break;
+      case '7': // Reserved
+        break;
+      case 'g': // Global PAX header
+        isHeaderFile = true;
+        this._globalPaxHeader = PaxHeader.parse(stream.readBuffer(size));
+        break;
+      case 'x': // PAX header
+        isHeaderFile = true;
+        paxHeader = PaxHeader.parse(stream.readBuffer(size));
+        break;
+      default:
+        // Unknown file type
+        break;
+    }
+
+    let dataEndPos = dataBeginPos + size;
+
+    // File data is padded to reach a 512 byte boundary; skip the padded bytes too.
+    if (size % 512 !== 0) {
+      dataEndPos += 512 - (size % 512);
+    }
+
+    stream.seekTo(dataEndPos);
+
+    const file = isHeaderFile
+      ? this._readNextFile()
+      : {
+          buffer,
+          name,
+          mode,
+          uid,
+          gid,
+          size,
+          mtime,
+          checksum,
+          type,
+          linkname,
+          ustarFormat,
+          version,
+          uname,
+          gname,
+          devmajor,
+          devminor,
+          namePrefix,
+        };
+
+    if (this._globalPaxHeader !== null) {
+      this._globalPaxHeader.applyHeader(file);
+    }
```

Comment:
I believe that for the first file the global pax header will be applied twice -- but that's probably OK as it's not additive.

Reply:
I re-jigged this a bit so that it's now only applied once.